### PR TITLE
Adds new look to the gamemode

### PR DESCRIPTION
**Pull request by Tadas3333**

### CODE REVIEW

File: gamemodes/bulletproof.pwn [hunk 1]
```diff
@@ -450,45 +469,9 @@ public ServerOnPlayerDeath(playerid, killerid, reason)
...
-					    PlayerTextDrawSetString(killerid, DeathText[killerid][0], sprintf("%sYou murdered %s~h~%s", MAIN_TEXT_COLOUR, TDC[Player[playerid][Team]], Player[playerid][Name]));
-					}
-					case 3:
-					{
-						PlayerTextDrawSetString(playerid, DeathText[playerid][1], sprintf("%s~h~%s%s sent you to cemetery", TDC[Player[killerid][Team]], Player[killerid][Name], MAIN_TEXT_COLOUR));
-                        PlayerTextDrawSetString(killerid, DeathText[killerid][0], sprintf("%sYou sent %s~h~%s%s to cemetery", MAIN_TEXT_COLOUR, TDC[Player[playerid][Team]], Player[playerid][Name], MAIN_TEXT_COLOUR));
-					}
-				}
-			}
-		}
```
Comment (khalidahmedshalabi) [inline gamemodes/bulletproof.pwn]: Why? I remember a few guys liked those messages and put them in their fragmovies as they sounded cool. Bring them back, please.

File: gamemodes/bulletproof.pwn [hunk 2]
```diff
@@ -9434,11 +9424,12 @@ public OnScriptUpdate()
...
 
-
 		// Update net info textdraws
 		if(PlayerInterface[i][INTERFACE_NET])
 		{
-  			PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS %s%d %sPing %s%d %sPacketLoss %s%.1f%%", MAIN_TEXT_COLOUR, TDC[Player[i][Team]], Player[i][FPS], MAIN_TEXT_COLOUR, TDC[Player[i][Team]], GetPlayerPing(i), MAIN_TEXT_COLOUR, TDC[Player[i][Team]], NetStats_PacketLossPercent(i)));
+  			//PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS %s%d %sPing %s%d %sPacketLoss %s%.1f%%", MAIN_TEXT_COLOUR, TDC[Player[i][Team]], Player[i][FPS], MAIN_TEXT_COLOUR, TDC[Player[i][Team]], GetPlayerPing(i), MAIN_TEXT_COLOUR, TDC[Player[i][Team]], NetStats_PacketLossPercent(i)));
+  			PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS ~r~%d %sPing ~r~%d %sPacketLoss ~r~%.1f%%", MAIN_TEXT_COLOUR, Player[i][FPS], MAIN_TEXT_COLOUR, GetPlayerPing(i), MAIN_TEXT_COLOUR, NetStats_PacketLossPercent(i)));
+            Update3DTextLabelText(Player[i][InfoLabel], -1, sprintf("%sPL: {FFFFFF}%.1f%%\n%sPing: {FFFFFF}%d\n%sFPS: {FFFFFF}%d", TextColor[Player[i][Team]], NetStats_PacketLossPercent(i), TextColor[Player[i][Team]], GetPlayerPing(i), TextColor[Player[i][Team]], Player[i][FPS]));
```
Comment (khalidahmedshalabi) [inline gamemodes/bulletproof.pwn]: I removed 3D text labels attached to players because they usually messed up. You would usually see player-A's label on player-B's body. Anyway, let me ask, are they highly necessary? Don't we already have target-info textdraws for the same purpose?

File: pawno/include/modules/src/gunmenu.inc [hunk 3]
```diff
@@ -198,47 +198,73 @@ ShowGunmenuDialog(playerid)
...
+					    else format(str, sizeof str, "%s\n{FFFFFF}%d\t{7777FF}%s\t{7777FF}%d", str, i, WeaponNames[GunmenuData[i][GunID]], GunmenuData[i][TimesSelected][team]);
+					}
 				}
 			}
-			ShowPlayerDialog(playerid, DIALOG_GUNMENU, DIALOG_STYLE_TABLIST_HEADERS, sprintf("%sGunmenu", TextColor[team]), str, sprintf("%sPick", TextColor[team]), sprintf("%sReset", TextColor[team]));
+			
+			ShowPlayerDialog(playerid, DIALOG_GUNMENU, DIALOG_STYLE_TABLIST_HEADERS, "{FFFFFF}Select your weapons:", str, "Pick", "Reset");
 		}
 	}
 }
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/gunmenu.inc:270]: Could you explain the need of the extra tab (ID)?

File: pawno/include/modules/src/match.inc [hunk 4]
```diff
@@ -1035,13 +1039,17 @@ public PauseStopWatch()
...
 	if(PauseTimeElapsed[1] < 1)
 		format(str, sizeof str, "~b~~h~Round has been paused for: %02d:%02d~n~~b~~h~~h~Pauses count so far: %d", PauseTimeElapsed[1], PauseTimeElapsed[0], TotalPausesInRound);
 	else if(PauseTimeElapsed[1] == 1)
 		format(str, sizeof str, "~y~~h~Round has been paused for: %02d:%02d~n~~y~~h~~h~Pauses count so far: %d", PauseTimeElapsed[1], PauseTimeElapsed[0], TotalPausesInRound);
 	else
 		format(str, sizeof str, "~r~~h~Round has been paused for: %02d:%02d~n~~r~~h~~h~Pauses count so far: %d", PauseTimeElapsed[1], PauseTimeElapsed[0], TotalPausesInRound);
-
+	*/
+    format(str, sizeof str, "~r~~h~Round is paused!~n~~w~Time: ~r~~h~%02d:%02d", PauseTimeElapsed[1], PauseTimeElapsed[0]);
+
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/match.inc]: Why did you remove the count of total pauses in a round?

File: pawno/include/modules/src/spectate.inc [hunk 5]
```diff
@@ -140,11 +140,18 @@ UpdatePlayerSpectators(playerid)
...
 		foreach(new i : PlayerSpectators[playerid]) // Loop through the array that contains this player's spectators
 		{
+		    // Make sure we don't show more players than MAX_PLAYER_SPECTATORS. Otherwise string will go beyond 255th character and cause crashes.
+		    if(totalAdded >= MAX_PLAYER_SPECTATORS)break;
+		
 			// Format string with spectator name
-			format(str, sizeof str, "%s%s~n~", str, Player[i][NameWithoutTag]);
+			format(str, sizeof str, "%s%s ~r~~h~%d~n~", str, Player[i][NameWithoutTag], i);
+			
+			totalAdded++;
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/spectate.inc:151]: This is not needed at all. Limit checks were already done in the lines below (164 and 389, for example) which you removed (and I don't understand why). 

``` PAWN
if(Iter_Count(PlayerSpectators[i]) >= MAX_PLAYER_SPECTATORS)
 	continue; // Skip this player if he's got enough people spectating him
```

File: gamemodes/bulletproof.pwn [hunk 6]
```diff
@@ -9434,11 +9424,12 @@ public OnScriptUpdate()
 	    // Show target info
 	    ShowTargetInfo(i, GetPlayerTargetPlayer(i));
 
-
 		// Update net info textdraws
 		if(PlayerInterface[i][INTERFACE_NET])
 		{
-  			PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS %s%d %sPing %s%d %sPacketLoss %s%.1f%%", MAIN_TEXT_COLOUR, TDC[Player[i][Team]], Player[i][FPS], MAIN_TEXT_COLOUR, TDC[Player[i][Team]], GetPlayerPing(i), MAIN_TEXT_COLOUR, TDC[Player[i][Team]], NetStats_PacketLossPercent(i)));
+  			//PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS %s%d %sPing %s%d %sPacketLoss %s%.1f%%", MAIN_TEXT_COLOUR, TDC[Player[i][Team]], Player[i][FPS], MAIN_TEXT_COLOUR, TDC[Player[i][Team]], GetPlayerPing(i), MAIN_TEXT_COLOUR, TDC[Player[i][Team]], NetStats_PacketLossPercent(i)));
+  			PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS ~r~%d %sPing ~r~%d %sPacketLoss ~r~%.1f%%", MAIN_TEXT_COLOUR, Player[i][FPS], MAIN_TEXT_COLOUR, GetPlayerPing(i), MAIN_TEXT_COLOUR, NetStats_PacketLossPercent(i)));
```
Comment (khalidahmedshalabi) [inline gamemodes/bulletproof.pwn]: Is that style so hated? I'm talking about the style that your ping/kill/death textdraws color change to the color of your team (red/blue/yellow or whatever).

File: pawno/include/modules/src/textdraws.inc [hunk 7]
```diff
@@ -1030,6 +1038,17 @@ LoadPlayerTextDraws(playerid)
 	return 1;
 }
 
+UpdateRoundKillDmgTDmg(playerid)
+{
+    new iString[160];
+	
+	//format(iString, sizeof(iString), "%sKills %s%d~n~%sDamage %s%d~n~%sTotal Dmg %s%d", MAIN_TEXT_COLOUR, TDC[Player[playerid][Team]], Player[playerid][RoundKills], MAIN_TEXT_COLOUR, TDC[Player[playerid][Team]], Player[playerid][RoundDamage], MAIN_TEXT_COLOUR, TDC[Player[playerid][Team]], Player[playerid][TotalDamage]);
+	format(iString, sizeof(iString), "%sKills ~r~%d~n~%sDamage ~r~%d~n~%sTotal Dmg ~r~%d", MAIN_TEXT_COLOUR,Player[playerid][RoundKills], MAIN_TEXT_COLOUR, Player[playerid][RoundDamage], MAIN_TEXT_COLOUR, Player[playerid][TotalDamage]);
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/textdraws.inc:1129]: Again, I'm asking, is this style so hated that you had to change it? 😞

File: pawno/include/modules/src/textdraws.inc [hunk 8]
```diff
@@ -1387,16 +1409,22 @@ public DoShowIntroTextDraws(index, response_code, data[])
...
 ShowIntroTextDraws(playerid)
 {
+    #pragma unused playerid
+    /*
 	HTTP(playerid, HTTP_GET, "infinite-gaming.ml/khk/bulletproof/league/news_textdraw_text.php", "", "DoShowIntroTextDraws");
 
 	TextDrawShowForPlayer(playerid, Intro_GoWebsite);
 	TextDrawShowForPlayer(playerid, Intro_GetStarted);
 	TextDrawShowForPlayer(playerid, Intro_News);
+    */
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/textdraws.inc:1547]: The intro textdraws should be redesigned then? Because they're needed for league.

File: pawno/include/modules/src/teaminfotd.inc [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+	}
+    return 1;
+}
+
+UpdateTeamCornerInfo()
+{
+	if(ShowHPBars)
+		return 0;
+	
+    // Variables for sorting
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/teaminfotd.inc]: So man I've always wondered why you and whoever wrote the same thing for attdef liked sorting these info so much? Do they really need to be sorted in this place? I mean, you already get them sorted in match results at the end of the round and I never needed to see them sorted while PLAYING the round. The fact that this function is called very often and you have to do all that processing over and over really gets on my nerves 😞

**Review by khalidahmedshalabi — CHANGES_REQUESTED** [review]

**Comment by beijind** [comment]
I am trying to make everything look more familiar with Attack-Defend gamemode. Stats on the left and net stats on the top must be white-red colored. And I am trying to make gunmenu look similar also.

Your spectators limit didn't let people spectate other players because limit was reached (Wtf is this?). 
Why the hell I can't spectate people? The only thing why limit is needed for is the textdraw thing which I took care of.

I look at text labels even more than on the left side textdraw, also labels are closer to the crosshair so it is easier for my eyes,  and I don't know what technique you used for those labels, I attach them when player connects, so it shouldn't get fucked up I guess (like on Art of War gm).

That textdraw must be sorted otherwise it doesn't look right. Looks worse than on Attack-Defend then, unprofessional, and that function is not called every second, so you can chill. Tho quick sort algorithm could be implemented instead of bubble sort.

Those kill messages looked childish for me, idk (owned and shit), could make them activatable in /config. 

I don't think many people like that league system. People were happy that league was removed in my server.

File: gamemodes/bulletproof.pwn (hunk 1 above)
Comment (MiGu3X) [inline gamemodes/bulletproof.pwn]: You should add an option for users to toggle death messages.

That may be better since anyone can have the feature they want.

Comment (khalidahmedshalabi) [reply gamemodes/bulletproof.pwn]: Okay

File: gamemodes/bulletproof.pwn (hunk 6 above)
Comment (Tadas3333) [reply gamemodes/bulletproof.pwn]: I don't like it, I prefer red/white combination and I think most people liked it on attack-defend gamemode.

File: gamemodes/bulletproof.pwn (hunk 2 above)
Comment (Tadas3333) [reply gamemodes/bulletproof.pwn]: Labels are always closer to the crosshair so for me personally it's easier. I don't really look at the left textdraw because I keep on aiming all the time. I need to focus on the target. And I look at the left textdraw when player is far away and I can't see the label. So I think both are needed.

I implemented Art of War gamemode system, that labels get attached when player joins, and destroyed when player leaves, so labels can't get fucked up I suppose.

File: pawno/include/modules/src/gunmenu.inc (hunk 3 above)
Comment (Tadas3333) [reply pawno/include/modules/src/gunmenu.inc:270]: Attack-Defend gamemode has IDs, and I try to make everything familiar including gunmenu.

File: pawno/include/modules/src/spectate.inc (hunk 5 above)
Comment (Tadas3333) [reply pawno/include/modules/src/spectate.inc:151]: Your limit didn't let people spectate others because limit was reached (Wtf is that?) Why I can't spectate other people? The only reason why limit is needed is the textdraw that I took care of.

File: pawno/include/modules/src/textdraws.inc (hunk 7 above)
Comment (Tadas3333) [reply pawno/include/modules/src/textdraws.inc:1129]: I mean blue looks too weird when everyone are used to red and white.

File: pawno/include/modules/src/textdraws.inc (hunk 8 above)
Comment (Tadas3333) [reply pawno/include/modules/src/textdraws.inc:1547]: I don't think league should be enabled by default as it's not very popular. And people congratulated me for removing league system in my server.

File: gamemodes/bulletproof.pwn (hunk 6 above)
Comment (khalidahmedshalabi) [reply gamemodes/bulletproof.pwn]: Alright, let's give people what they like.

File: gamemodes/bulletproof.pwn (hunk 2 above)
Comment (khalidahmedshalabi) [reply gamemodes/bulletproof.pwn]: If they work well then okay leave them, I just thought that you copied the code from attdef.

File: pawno/include/modules/src/gunmenu.inc (hunk 3 above)
Comment (khalidahmedshalabi) [reply pawno/include/modules/src/gunmenu.inc:270]: Nah I don't think this is necessary. Not much notable, I mean, so please remove it.

File: pawno/include/modules/src/spectate.inc (hunk 5 above)
Comment (khalidahmedshalabi) [reply pawno/include/modules/src/spectate.inc:151]: Then just increase the limit MAX_PLAYER_SPECTATORS in modules/header/spectate?

File: pawno/include/modules/src/textdraws.inc (hunk 7 above)
Comment (khalidahmedshalabi) [reply pawno/include/modules/src/textdraws.inc:1129]: Alright.

File: pawno/include/modules/src/textdraws.inc (hunk 8 above)
Comment (khalidahmedshalabi) [reply pawno/include/modules/src/textdraws.inc:1547]: Okay, leave this part for me. I'll rework some stuff in league system.

File: gamemodes/attdef.pwn [hunk 10]
```diff
@@ -450,45 +537,54 @@ public ServerOnPlayerDeath(playerid, killerid, reason)
 	}
 	else
 	{
-		switch(reason)
+		if(!RandomDeathMessages)
```
Comment (khalidahmedshalabi) [inline gamemodes/attdef.pwn:537]: You said people already complained about those messages being inappropriate, I think we should completely remove them then.

File: gamemodes/attdef.pwn [hunk 11]
```diff
@@ -13,6 +13,10 @@
 #include <sampac> 			// THE MIGHTY NEW ANTICHEAT
 #include <mSelection>       // Selection with preview models feature library
 #include <gBugFix>			// Fix false vehicle entry as passenger (G (teleport/distance) bug)
+#include <md-sort>          // Sorts multi dimensional arrays
+
+#define RELEASE_VERSION 1
+#define ROUND_NEVER_END 0   // Used for testing purposes only
```
Comment (khalidahmedshalabi) [inline gamemodes/attdef.pwn]: I don't see a reason for this to be here in main script, maybe a better place could be in header/match.txt

File: gamemodes/attdef.pwn [hunk 12]
```diff
@@ -166,24 +191,76 @@ public OnPlayerRequestClass(playerid, classid)
 	SetPlayerTime(playerid, 12, 0);
 	SetPlayerInterior(playerid, MainInterior);
 	
-	switch(random(3))
+	if(!SightseeingInClassSelection)
```
Comment (khalidahmedshalabi) [inline gamemodes/attdef.pwn:192]: What does this do and can I toggle it on and off via /config?

**Review by khalidahmedshalabi — CHANGES_REQUESTED** [review]

File: pawno/include/modules/src/spectate.inc (hunk 5 above)
Comment (Tadas3333) [reply pawno/include/modules/src/spectate.inc:151]: Then why there is even a limit? Why is it not unlimited like on other gamemodes?

File: gamemodes/attdef.pwn (hunk 10 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn:537]: I made it toggleable in /config and disabled by default. So there is no problem I suppose as it will satisfy everyone.
More configurable options will make more unique gamemodes.

File: gamemodes/attdef.pwn (hunk 12 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn:192]: Yes, it is disabled by default. Normally when you are in a class selection, camera moves around the spawn. With this enabled, camera goes to random places around San Andreas map.

File: gamemodes/attdef.pwn (hunk 11 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn]: It's easy switch this before pressing compile button, but if you don't like it then sure.

File: gamemodes/bulletproof.pwn (hunk 2 above)
Comment (Tadas3333) [reply gamemodes/bulletproof.pwn]: I don't have source code of attdef at all.

Comment (khalidahmedshalabi) [reply gamemodes/bulletproof.pwn]: You're lucky you don't 😄

File: pawno/include/modules/src/spectate.inc (hunk 5 above)
Comment (khalidahmedshalabi) [reply pawno/include/modules/src/spectate.inc:151]: I really don't remember. I guess I did it for textdraws as you said, so you're right.

File: gamemodes/attdef.pwn (hunk 11 above)
Comment (khalidahmedshalabi) [reply gamemodes/attdef.pwn]: I'm talking about where you put it. It's related to nothing but match mode, so IMO header/match is the most suitable place. One of my goals: https://github.com/KHKKhalid/SAMPBulletproof/issues/8

File: gamemodes/attdef.pwn (hunk 12 above)
Comment (khalidahmedshalabi) [reply gamemodes/attdef.pwn:192]: Like totally random places or preset by you?

File: pawno/include/modules/src/weaponshot.inc [hunk 13]
```diff
@@ -597,6 +600,19 @@ CheckVisualDamageTextDraws()
 				Iter_SafeRemove(PlayerHasTVisualDmg, cur, i); // Safely remove him from the iterator array.
 		}
 	}
+	
+    // Update Damage Bubble
+	foreach(new i : Player)
```
Comment (khalidahmedshalabi) [inline pawno/include/modules/src/weaponshot.inc]: I really prefer to use y_iterate for such operations instead of a regular foreach over all players with an if statement.

**Review by khalidahmedshalabi — CHANGES_REQUESTED** [review]

File: gamemodes/bulletproof.pwn [hunk 14]
```diff
@@ -13,6 +13,7 @@
 #include <sampac> 			// THE MIGHTY NEW ANTICHEAT
 #include <mSelection>       // Selection with preview models feature library
 #include <gBugFix>			// Fix false vehicle entry as passenger (G (teleport/distance) bug)
+#include <md-sort>         // Sorts multi dimensional arrays
```
Comment (KHKKhalid) [inline gamemodes/bulletproof.pwn]: Maybe we should use this to sort players results at the end of the match too in the future.

File: gamemodes/attdef.pwn [hunk 15]
```diff
@@ -13,6 +13,10 @@
 #include <sampac> 			// THE MIGHTY NEW ANTICHEAT
 #include <mSelection>       // Selection with preview models feature library
 #include <gBugFix>			// Fix false vehicle entry as passenger (G (teleport/distance) bug)
+#include <md-sort>          // Sorts multi dimensional arrays
```
Comment (khalidahmedshalabi) [inline gamemodes/attdef.pwn:17]: Maybe we should use this to sort players results at the end of the match too in the future.

File: gamemodes/attdef.pwn [hunk 16]
```diff
@@ -9634,6 +9644,16 @@ public OnScriptUpdate()
   			PlayerTextDrawSetString(i, FPSPingPacket[i], sprintf("%sFPS ~r~%d %sPing ~r~%d %sPacketLoss ~r~%.1f%%", MAIN_TEXT_COLOUR, Player[i][FPS], MAIN_TEXT_COLOUR, GetPlayerPing(i), MAIN_TEXT_COLOUR, NetStats_PacketLossPercent(i)));
             Update3DTextLabelText(Player[i][InfoLabel], -1, sprintf("%sPL: {FFFFFF}%.1f%%\n%sPing: {FFFFFF}%d\n%sFPS: {FFFFFF}%d", TextColor[Player[i][Team]], NetStats_PacketLossPercent(i), TextColor[Player[i][Team]], GetPlayerPing(i), TextColor[Player[i][Team]], Player[i][FPS]));
 		}
+		
+		// Update Speedometer
```
Comment (khalidahmedshalabi) [inline gamemodes/attdef.pwn]: Nah man remove that shit from the gamemode. It's about murder here.

File: gamemodes/attdef.pwn (hunk 12 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn:192]: By me, look at the code.

File: gamemodes/attdef.pwn (hunk 15 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn:17]: Maybe maybe

File: gamemodes/attdef.pwn (hunk 16 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn]: The point of speedometer is not the speed, but vehicle health. I constantly need to write /dl to see vehicles health, this could be toggleable in /config.

Comment (Tadas3333) [reply gamemodes/attdef.pwn]: It is very useful for me and didn't hear any complaints.

File: pawno/include/modules/src/weaponshot.inc (hunk 13 above)
Comment (Tadas3333) [reply pawno/include/modules/src/weaponshot.inc]: Ok, I actually never used anything with iteration other than for all players. Will try to implement that.

File: gamemodes/attdef.pwn (hunk 16 above)
Comment (Tadas3333) [reply gamemodes/attdef.pwn]: Tho I could make it show health only if you want.

Comment (khalidahmedshalabi) [reply gamemodes/attdef.pwn]: If it's for that purpose then just make a /vh command (show vehicle's hp percentage and damage status in a client message). It's short enough to type it quickly and you don't have to worry about toggling it on and off. I'm very sensitive to stuff under OnScriptUpdate and OnPlayerUpdate. You shouldn't add code there unless you're in a really bad need.

Comment (Tadas3333) [reply gamemodes/attdef.pwn]: No, command is fucking stupid, i need to keep driving or piloting to avoid bullets. And servers won't die if it's there.

Comment (Tadas3333) [reply gamemodes/attdef.pwn]: I would better use /dl then.
